### PR TITLE
Upload documents in Materials

### PR DESCRIPTION
**Pull request by FlorianDinant**
# Upload Documents in Materials

Hello ! This pull request introduces a new feature that allows users (more specifically the teacher) to upload documents to individual Materials. It enhances the application’s functionality by enabling seamless file management and organization.

I noticed that some changes have been added in this PR that I have not written (`community.tsx` / `horizontalCalendar.tsx`...)

### Upload Documents to Materials :

- Users can now select and upload documents to specific Materials.
- Supported file formats include PDF, images (PNG, JPG, JPEG, WEBP, GIF etc.), and other common types.
- Files are saved directly into a directory dedicated to the application.

### DocumentDisplay.tsx
Now, to provide more readability and reusability, the documents are displayed via the new component `DocumentDisplay`.
It allows to make the document deletable or not according to whether the user is a teacher and is on the edit material page or not.

## Error Handling and User Feedback:
- Success or failure messages are displayed appropriately to improve user experience.
- An alert is displayed to ask confirmation or cancellation when the user wants to delete the material to make sure it is not a miss click.
- An alert is displayed when the user wants to update the changes but some fields are not empty is the `Add Document` section and ask for confirmation.

## Other useful informations
- The firebase cloud storage now don't store the files directly in the cloud root directory, it organise the files according to the course and the material. exemple: `courses/edweiss-demo/materials/tMbRhyfV8K4nDkAYY45g/relativity.pdf`
- 3 new firebase functions have been added say : 
   - `uploadToFirebase` : allows to add a new material on firebase cloud storage
   - `deleteFromFirebase` : allows to delete a material from firebase cloud storage
   - `deleteDirectoryFromFirebase` : allows to delete a whole directory in the firebase cloud storage when the user delete the entire material

## Testing
- Test edge cases
- `MaterialComponent.text.tsx` have been updated with 90% coverage to handle the new added code

## Screenshots
Screenshot 2024-12-17 at 12 26 59
Screenshot 2024-12-17 at 12 27 13
Screenshot 2024-12-17 at 12 29 18
Screenshot 2024-12-17 at 12 29 50
Screenshot 2024-12-17 at 12 30 31
Screenshot 2024-12-17 at 12 45 34

Let me know if you need further adjustments!

### CODE REVIEW

**Review by TuanDangNguyen2003 — APPROVED** [review]
Thank you for the detailed PR description and for the effort you’ve put into this! The feature to upload documents to individual materials is a significant enhancement to the app's functionality, particularly for teachers.  

---
1. **UI and UX**:  
   The UI looks clean and intuitive, based on the screenshots. Alerts for deletion and field validation are a great touch for preventing accidental actions and improving user experience.  

2. **Code Organization and Maintainability**:  
   - The creation of the `DocumentDisplay` component is an excellent decision. Separating the document display logic makes the code more modular, reusable, and easier to maintain in the future.  
   - The Firebase directory structure is well-organized by course and material, which avoids clutter in the cloud storage and ensures better file management.  

3. **Testing**:  
   It’s great to see you have thoroughly tested the new functionality, achieving **>90% coverage** in `MaterialComponent.test.tsx`. Testing edge cases shows a strong focus on reliability.  

4. **Essential Functionalities**:  
   - The addition of Firebase functions (`uploadToFirebase`, `deleteFromFirebase`, `deleteDirectoryFromFirebase`) ensures smooth handling of file uploads, deletions, and storage organization.  
   - Error handling and user feedback are thoughtfully implemented, making the feature robust and user-friendly.  

---

### **Minor Observations**  
- You mentioned that some changes were added in `community.tsx` and `horizontalCalendar.tsx` that you did not write. If these are unrelated to this feature, it might be good to clean them up or confirm why they’re included.  
- Consider adding a quick note in the PR about performance or file size restrictions, if any, for uploads. This can help in setting user expectations.  

---

### **Summary**  
This is an excellent PR with a well-thought-out design, clean modular code, and strong testing coverage. The functionality is essential, and you’ve ensured maintainability with a good separation of concerns (e.g., `DocumentDisplay`). Overall, great work!  

Let me know if you need further feedback, but this looks ready to merge. 🚀

File: edweiss-app/app/(app)/courses/[id]/materials/[materialId]/index.tsx [hunk 1]
```diff
@@ -261,7 +263,7 @@ const DocumentScreen: ApplicationRoute = () => {
     // ------------------------------------------------------------
     const displayPageNumber = () => (
         <TView testID={testIDs.pageNumView} backgroundColor='transparent' radius={5} px={10} py={5} style={{ position: 'absolute', top: 10, left: 10, backgroundColor: 'rgba(128, 128, 128, 0.2)' }}>
-            <TText testID={testIDs.pageNumText} bold color='text'>
+            <TText testID={testIDs.pageNumText} bold color='constantBlack'>
```
[the remark this answers is not in the document]
Comment (FlorianDinant) [reply edweiss-app/app/(app)/courses/[id]/materials/[materialId]/index.tsx:266]: Yes actually it looks good because the display number in placed on the PDF. The pages of PDFs are white so even in dark mode, we have to display in black. Do you see what I mean ?

File: edweiss-app/components/core/SVGImage.tsx [hunk 2]
```diff
@@ -25,12 +25,11 @@ const SVGImage: ReactComponent<SVGImageProps> = ({ uri, width, height, color, ti
...
                 const response = await fetch(uri);
-                const svgText = await response.text();
-                setSvgData(svgText);
-                setLoading(false);
+                const svgData = await response.text();
+                setSvgData(svgData);
             } catch (error) {
                 console.error('Error fetching SVG: ', error);
-                console.debug(svgData);
+            } finally {
```
[the remark this answers is not in the document]
Comment (FlorianDinant) [reply edweiss-app/components/core/SVGImage.tsx:32]: Finally executes what you give inside in both cases
This means that this part of code is executed event if an error is raised

File: edweiss-app/app/(app)/courses/[id]/index.tsx [hunk 3]
```diff
@@ -313,7 +332,7 @@ const CoursePage: ApplicationRoute = () => {
 				</For>
 
 				{/* Bouton vers les Passed Assignments */}
```
Comment (FerdinandGomerChiu) [inline edweiss-app/app/(app)/courses/[id]/index.tsx:334]: A nitpick: comments should be in english

File: edweiss-app/app/(app)/courses/[id]/index.tsx [hunk 4]
```diff
@@ -370,8 +392,9 @@ const CoursePage: ApplicationRoute = () => {
 				animationType="fade"
 				transparent={true}
 				onRequestClose={() => setActionModalVisible(false)}
+				testID='actionModal'
 			>
-				<Animated.View style={{ flex: 1, justifyContent: 'center', alignItems: 'center', backgroundColor: 'rgba(0, 0, 0, 0.5)', opacity: fadeAnim }}>
+				<Animated.View testID='animatedViewActionSelection' style={{ flex: 1, justifyContent: 'center', alignItems: 'center', backgroundColor: 'rgba(0, 0, 0, 0.5)', opacity: fadeAnim }}>
 					<SelectActions onOutsideClick={closeModalOnOutsideClick} onSelectAssignment={() => handleActionSelect('addAssignment')} onSelectMaterial={() => handleActionSelect('addMaterial')} />
 				</Animated.View>
 			</Modal>
```
Comment (FerdinandGomerChiu) [inline edweiss-app/app/(app)/courses/[id]/index.tsx:400]: Using Modal  works, but you could have also used a different type of modal (ModalContainerView), which most of us used already.

File: edweiss-app/components/courses/MaterialComponent.tsx [hunk 5]
```diff
@@ -87,63 +143,98 @@ interface MaterialProps {
...
+        setDocTitleChanged(false);
+    }
 
+    //===============================================================
+    //======================== Listeners ============================
+    //===============================================================
+
+    // Document Format Listener
+    useEffect(() => {
+        if (docName) {
```
Comment (FerdinandGomerChiu) [inline edweiss-app/components/courses/MaterialComponent.tsx:196]: you wrote if(string) : I do not understand this syntax, maybe consider putting a boolean instead.

File: edweiss-app/components/courses/MaterialComponent.tsx [hunk 6]
```diff
@@ -177,30 +265,191 @@ const MaterialComponent: ReactComponent<MaterialProps> = ({ mode, onSubmit, onDe
...
 
+
+
+    //===============================================================
+    //======================== File Upload ==========================
+    //===============================================================
+
+    // Handle the file picker
+    const handlePickFile = async () => {
+        try {
```
Comment (FerdinandGomerChiu) [inline edweiss-app/components/courses/MaterialComponent.tsx:276]: You define a lot of functions inside your Material component. I see you still got sufficient coverage (excellent job on that), but if you want to increase it further, consider separating the functions from the component, to be able to test them easily.

**Review by FerdinandGomerChiu — APPROVED** [review]
This is significant, quality work. I am also impressed by the coverage you got, at the same time you are making changes. Keep it up!

File: edweiss-app/app/(app)/courses/[id]/index.tsx (hunk 4 above)
Comment (FlorianDinant) [reply edweiss-app/app/(app)/courses/[id]/index.tsx:400]: The Modal that I use, has a different behaviour than ModalContainerView

File: edweiss-app/app/(app)/courses/[id]/index.tsx (hunk 3 above)
Comment (FlorianDinant) [reply edweiss-app/app/(app)/courses/[id]/index.tsx:334]: Yes! Thank you

File: edweiss-app/components/courses/MaterialComponent.tsx (hunk 5 above)
Comment (FlorianDinant) [reply edweiss-app/components/courses/MaterialComponent.tsx:196]: If docName is empty, null or undefined it will not be evaluated as true. That is why I write it like this

File: edweiss-app/components/courses/MaterialComponent.tsx (hunk 6 above)
Comment (FlorianDinant) [reply edweiss-app/components/courses/MaterialComponent.tsx:276]: Yes this is a good idea! I will do it if possible (depends if I have time to do it) in a future PR.
